File: cores/csharp/src/Audio/FdsAudio.cs
using System.Runtime.CompilerServices;

namespace NesCore;

// Famicom Disk System expansion audio (wavetable + modulator).
// Port of src/mappers/fds_audio.rs.
public sealed class FdsAudio
{
    public const int WaveTableSize = 64;
    public const int ModTableSize = 64;

    public byte[] WaveRam = new byte[WaveTableSize];
    public byte WaveAddr;
    public byte[] ModWave = new byte[ModTableSize];
    public byte ModWaveAddr;
    public byte MasterVolume;
    public bool EnvDisabled;
    public bool EnvIncrease;
    public byte VolumeGain;
    public ushort Freq;
    public ushort ModFreq;
    public bool ModDisabled = true;
    public byte ModSweepNeg;
    public byte ModSweepPos;
    public sbyte ModGain;
    public byte ModGainOutput;
    public byte ModSweepCounter;
    public uint PhaseAcc;
    public uint ModPhaseAcc;

    public FdsAudio() { }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static sbyte SatSubI8(sbyte a, sbyte b)
    {
        int r = a - b;
        if (r < -128) return -128;
        if (r > 127) return 127;
        return (sbyte)r;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static sbyte SatAddI8(sbyte a, sbyte b)
    {
        int r = a + b;
        if (r < -128) return -128;
        if (r > 127) return 127;
        return (sbyte)r;
    }

    public void WriteRegister(ushort addr, byte value)
    {
        if (addr >= 0x4040 && addr <= 0x407F)
        {
            WaveRam[WaveAddr % WaveTableSize] = (byte)(value & 0x3F);
            WaveAddr = (byte)((WaveAddr + 1) & 0x3F);
            return;
        }
        switch (addr)
        {
            case 0x4080:
                MasterVolume = (byte)(value & 0x3F);
                EnvDisabled = (value & 0x40) != 0;
                EnvIncrease = (value & 0x80) != 0;
                if (EnvDisabled) VolumeGain = MasterVolume;
                else VolumeGain = EnvIncrease ? (byte)0 : (byte)0x3F;
                break;
            case 0x4081:
                Freq = (ushort)((Freq & 0x00FF) | ((value & 0x0F) << 8));
                break;
            case 0x4082:
                Freq = (ushort)((Freq & 0x0F00) | value);
                break;
            case 0x4083:
                ModFreq = (ushort)((ModFreq & 0x00FF) | ((value & 0x0F) << 8));
                if ((value & 0x80) != 0) { ModDisabled = true; ModPhaseAcc = 0; }
                break;
            case 0x4084:
                ModSweepNeg = (byte)(value & 0x0F);
                break;
            case 0x4085:
                ModSweepPos = (byte)(value & 0x0F);
                break;
            case 0x4086:
                ModGain = SatSubI8(ModGain, (sbyte)(value & 0x3F));
                break;
            case 0x4087:
                ModGain = SatAddI8(ModGain, (sbyte)(value & 0x3F));
                if ((value & 0x80) != 0) ModDisabled = true;
                break;
            case 0x4088:
                ModWave[ModWaveAddr % ModTableSize] = (byte)(value & 0x07);
                ModWaveAddr = (byte)((ModWaveAddr + 1) & 0x3F);
                break;
            case 0x4089:
                MasterVolume = (byte)((MasterVolume & 0x3C) | (value & 0x03));
                break;
            case 0x408A:
                if ((value & 0x80) != 0) ModDisabled = true;
                break;
        }
    }

    public byte ReadRegister(ushort addr)
    {
        switch (addr)
        {
            case 0x4090: return (byte)(VolumeGain & 0x3F);
            case 0x4092: return (byte)(ModGainOutput & 0x3F);
            default: return 0;
        }
    }

    public void Clock(uint apuCycles)
    {
        for (uint c = 0; c < apuCycles; ++c)
        {
            PhaseAcc += Freq;
            if (!ModDisabled)
            {
                ModPhaseAcc += ModFreq;
                ModSweepCounter++;
                if (ModSweepCounter >= 8)
                {
                    ModSweepCounter = 0;
                    if (ModSweepNeg > 0)
                        ModGain = SatSubI8(ModGain, (sbyte)ModSweepNeg);
                    if (ModSweepPos > 0)
                        ModGain = SatAddI8(ModGain, (sbyte)ModSweepPos);
                }
            }
            if (!EnvDisabled && ModSweepCounter == 0)
            {
                if (EnvIncrease && VolumeGain < 0x3F) VolumeGain++;
                else if (!EnvIncrease && VolumeGain > 0) VolumeGain--;
            }
        }
    }

    public float Sample()
    {
        if (Freq == 0) return 0.0f;
        uint waveIdx = (PhaseAcc >> 12) & 0x3F;
        short modOffset = 0;
        if (!ModDisabled)
        {
            uint modIdx = (ModPhaseAcc >> 12) & 0x3F;
            short modVal = (short)ModWave[modIdx];
            modOffset = (short)((ModGain * modVal) >> 3);
        }
        short effIdx = (short)((waveIdx + modOffset) & 0x3F);
        short waveVal = (short)WaveRam[effIdx];
        short vol = (short)VolumeGain;
        short sample = (short)((waveVal - 32) * vol);
        float normalized = sample / 2016.0f;
        if (normalized < -1.0f) normalized = -1.0f;
        if (normalized > 1.0f) normalized = 1.0f;
        return normalized;
    }
}

File: cores/typescript/src/audio/fds_audio.ts
// Famicom Disk System expansion audio (wavetable + modulator).
// Port of cores/c/src/fds_audio.c.

export const FDS_WAVE_TABLE_SIZE = 64;
export const FDS_MOD_TABLE_SIZE = 64;

function satSubI8(a: number, b: number): number {
    let r = a - b;
    if (r < -128) r = -128;
    if (r > 127) r = 127;
    return r;
}
function satAddI8(a: number, b: number): number {
    let r = a + b;
    if (r < -128) r = -128;
    if (r > 127) r = 127;
    return r;
}

export class FdsAudio {
    waveRam: Uint8Array = new Uint8Array(FDS_WAVE_TABLE_SIZE);
    waveAddr: number = 0;
    masterVolume: number = 0;
    envDisabled: boolean = false;
    envIncrease: boolean = false;
    volumeGain: number = 0;
    freq: number = 0;
    phaseAcc: number = 0;
    modFreq: number = 0;
    modPhaseAcc: number = 0;
    modWave: Uint8Array = new Uint8Array(FDS_MOD_TABLE_SIZE);
    modWaveAddr: number = 0;
    modDisabled: boolean = true;
    modGain: number = 0;
    modSweepCounter: number = 0;
    modSweepNeg: number = 0;
    modSweepPos: number = 0;
    modGainOutput: number = 0;

    init(): void {
        this.waveRam.fill(0);
        this.waveAddr = 0;
        this.masterVolume = 0;
        this.envDisabled = false;
        this.envIncrease = false;
        this.volumeGain = 0;
        this.freq = 0;
        this.phaseAcc = 0;
        this.modFreq = 0;
        this.modPhaseAcc = 0;
        this.modWave.fill(0);
        this.modWaveAddr = 0;
        this.modDisabled = true;
        this.modGain = 0;
        this.modSweepCounter = 0;
        this.modSweepNeg = 0;
        this.modSweepPos = 0;
        this.modGainOutput = 0;
    }

    writeRegister(addr: number, value: number): void {
        if (addr >= 0x4040 && addr <= 0x407F) {
            this.waveRam[this.waveAddr % FDS_WAVE_TABLE_SIZE] = value & 0x3F;
            this.waveAddr = (this.waveAddr + 1) & 0x3F;
            return;
        }
        switch (addr) {
            case 0x4080:
                this.masterVolume = value & 0x3F;
                this.envDisabled = (value & 0x40) !== 0;
                this.envIncrease = (value & 0x80) !== 0;
                if (this.envDisabled) this.volumeGain = this.masterVolume;
                else this.volumeGain = this.envIncrease ? 0 : 0x3F;
                break;
            case 0x4081:
                this.freq = (this.freq & 0x00FF) | ((value & 0x0F) << 8);
                break;
            case 0x4082:
                this.freq = (this.freq & 0x0F00) | value;
                break;
            case 0x4083:
                this.modFreq = (this.modFreq & 0x00FF) | ((value & 0x0F) << 8);
                if (value & 0x80) { this.modDisabled = true; this.modPhaseAcc = 0; }
                break;
            case 0x4084:
                this.modSweepNeg = value & 0x0F;
                break;
            case 0x4085:
                this.modSweepPos = value & 0x0F;
                break;
            case 0x4086: {
                const neg = value & 0x3F;
                this.modGain = satSubI8(this.modGain, neg);
                break;
            }
            case 0x4087: {
                const pos = value & 0x3F;
                this.modGain = satAddI8(this.modGain, pos);
                if (value & 0x80) this.modDisabled = true;
                break;
            }
            case 0x4088:
                this.modWave[this.modWaveAddr % FDS_MOD_TABLE_SIZE] = value & 0x07;
                this.modWaveAddr = (this.modWaveAddr + 1) & 0x3F;
                break;
            case 0x4089:
                this.masterVolume = (this.masterVolume & 0x3C) | (value & 0x03);
                break;
            case 0x408A:
                if (value & 0x80) this.modDisabled = true;
                break;
        }
    }

    readRegister(addr: number): number {
        switch (addr) {
            case 0x4090: return this.volumeGain & 0x3F;
            case 0x4092: return this.modGainOutput & 0x3F;
            default: return 0;
        }
    }

    clock(apuCycles: number): void {
        for (let c = 0; c < apuCycles; ++c) {
            this.phaseAcc = (this.phaseAcc + this.freq) >>> 0;
            if (!this.modDisabled) {
                this.modPhaseAcc = (this.modPhaseAcc + this.modFreq) >>> 0;
                this.modSweepCounter = (this.modSweepCounter + 1) & 0xFF;
                if (this.modSweepCounter >= 8) {
                    this.modSweepCounter = 0;
                    if (this.modSweepNeg > 0) this.modGain = satSubI8(this.modGain, this.modSweepNeg);
                    if (this.modSweepPos > 0) this.modGain = satAddI8(this.modGain, this.modSweepPos);
                }
            }
            if (!this.envDisabled && this.modSweepCounter === 0) {
                if (this.envIncrease && this.volumeGain < 0x3F) this.volumeGain = (this.volumeGain + 1) & 0xFF;
                else if (!this.envIncrease && this.volumeGain > 0) this.volumeGain = (this.volumeGain - 1) & 0xFF;
            }
        }
    }

    sample(): number {
        if (this.freq === 0) return 0.0;
        const waveIdx = (this.phaseAcc >> 12) & 0x3F;
        let modOffset = 0;
        if (!this.modDisabled) {
            const modIdx = (this.modPhaseAcc >> 12) & 0x3F;
            const modVal = this.modWave[modIdx];
            modOffset = (this.modGain * modVal) >> 3;
        }
        const effIdx = (waveIdx + modOffset) & 0x3F;
        const waveVal = this.waveRam[effIdx];
        const vol = this.volumeGain;
        const sample = (waveVal - 32) * vol;
        let normalized = sample / 2016.0;
        if (normalized < -1.0) normalized = -1.0;
        if (normalized > 1.0) normalized = 1.0;
        return normalized;
    }
}
